### PR TITLE
feat: add C implementation for `math/base/special/spence`

### LINKED ISSUE #894: [RFC]: Add C implementation for `@stdlib/math/base/special/spence`
### Description

This RFC proposes adding C implementation for [`@stdlib/math/base/special/spence`](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/spence).

```c
double stdlib_base_spence( const double x );
```

### Related Issues

None.

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by gunjjoshi**
Resolves #894.

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for [`math/base/special/spence`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/@stdlib/math/base/special/spence).

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves a part of #649
-   fixes #894 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

**Comment by gunjjoshi** [comment]
While implementing this, a bit higher tolerance was required in the tests. I checked the implementation with that of `cephes`, but it was the same.

File: lib/node_modules/@stdlib/math/base/special/spence/src/main.c
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/spence/src/main.c]: Should be `@return` instead of `@returns` in C function documentation:

```suggestion
* @return     function value
```

**Review by Planeshifter — APPROVED** [review]
Looks good to me, will merge after CI checks have passed (updated `@returns` to `@return`  in C function documentation.)

File: lib/node_modules/@stdlib/math/base/special/spence/lib/native.js
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/spence/lib/native.js]:
```suggestion
```

This extended comment is not necessary in bindings files, as there is no visible implementation in this file.

File: lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js:91]: This test tolerance differs from the JavaScript tolerance. Is there a reason why given the implementation?

**Review by kgryte — CHANGES_REQUESTED** [review]
Overall, this is good; however, we should investigate the test tolerance discrepancy before merging.

**Comment by kgryte** [comment]
This PR also supersedes https://github.com/stdlib-js/stdlib/pull/968. Thus, once this PR is merged, the other PR should be closed.

File: lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js:91]: This was the reason why I mentioned about different tolerances required for `C` and `javascript` implementations. I checked the `C` implementation, both against the original `cephes` and our `javascript` implementations, but still couldn't find any difference.

Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js:91]: @gunjjoshi It looks like the primary potential culprit is `ln`. I suggest manually checking to see there whether results diverge for medium sized input values to Spence.

Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js:91]: In `ln`, there is a rounding off happening for the `C` results, unlike that of `javascript` ones.

For instance,
In `C`, `ln(0.3261523046092184)` gives `-1.120391`, whereas in `javascript`, it comes out to be `-1.12039081466284`.
Similarly, 
In `C`, `ln(0.8436498997995991)` gives `-0.170018`, whereas in `javascript`, it comes out to be `-0.17001768110790474`.

Should this be a thing of concern?

Comment (Planeshifter) [reply lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js:91]: @gunjjoshi We would want the C implementation to match the JavaScript one. In general, we want the implementations to yield equivalent results. In this case, results shouldn't be rounded / truncated in the `ln` C implementation. Would be great if you were able to track that down and fix in a separate PR! 🙏

Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js:91]: Make sure that, when you are logging C values, you specify the appropriate precision in `printf` statements. That rounding looks suspicious to me and not the result of `ln`, per se.

Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js:91]: When I printed the results in `C` with `%.ng`, where `n` is the number of digits to be displayed after the decimal point, I found that the results do match the `javascript` ones. So, the problem might be somewhere else, as `ln` yields the same results for both `C` and `javascript` implementations, for medium sized input values.

Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js:91]: @kgryte @Planeshifter The problem seems to be at [L137](https://github.com/stdlib-js/stdlib/pull/2298/files#diff-762c1402388106d6bc0fe6a7c2e39048be27d1972dcf6c623b120c9747654f7dR137). 

Both `C` and `javascript` produce different results. For instance,

When `x = 0.9558294589178357`, 
`C`: `polyval_a( w ) / polyval_b( w ): 1.0112649603618284`
`JavaScript`: `polyvalA( w ) / polyvalB( w ): 1.0112649603618282`

Observe the difference in the last digits here.

Should this be due to the difference in the internal division implementation in `C` and `javascript`, leading to a little difference in the precision? In that case we might need different tolerances.

Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js:91]: @gunjjoshi Thanks for tracking this down! Interesting that division would yield differences. It's also possible that the polynomial equations are slightly different.

Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js:91]: @kgryte I checked the polynomial equations, they are also yielding different results individually. But I am not sure if that is due to any difference in implementation or due to the way I'm trying to print the results, resulting in rounding off of the values. Currently, I'm using `%.16g`, for example, as a format specifier in `C` to print 16 digits after the decimal point. Is there any other way?

Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js:91]: Okay. I tracked down the issue. The discrepancy arises from floating-point operation fusing (see https://clang.llvm.org/docs/UsersManual.html#cmdoption-ffp-contract). If we disable fusing, we seem to get expected results.

In short, it is fine for us to have a different tolerance for this package, as there isn't much we can do to actually ensure the same results in the wild, so to speak. Depends on what compiler flags are set, rather than something to do with our implementation.

Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js:91]: Got it. So how do we detect this floating-point operation fusing? Like, in our future implementations of other functions too, this might arise, right?

Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js:91]: In this case, we can see the difference by setting the tolerance back to `2.0 * EPS` and then doing

```
$ CFLAGS="-ffp-contract=off" make install-node-addons NODE_ADDONS_PATTERN="math/base/special/spence"
$ make test TESTS_FILTER=".*/math/base/special/spence/.*"
```

and then

```
$ CFLAGS="-ffp-contract=on" make install-node-addons NODE_ADDONS_PATTERN="math/base/special/spence"
$ make test TESTS_FILTER=".*/math/base/special/spence/.*"
```

This has probably happened in the past; we just hadn't caught it until now. Realistically, it is unlikely to be a problem. Main thing is that we investigated and ruled out implementation bugs. So long as we are able to explain it, should be fine.

Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js:91]: I added a comment to the `test.native.js` file. I think we can consider this discussion resolved.

**Review by kgryte — APPROVED** [review]
LGTM. Thanks for working on this @gunjjoshi. I'll merge once CI passes.

File: lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/spence/test/test.native.js:91]: Great, thanks for the detailed explanation!

**Pull request merged**
